tools/socket_passing.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python
+
+# This tool is a helper script that queries the admin address for all listener
+# addresses after envoy startup. (The admin adress is written out to a file by
+# setting the -a flag in the envoy binary.) The script then outputs a new json
+# config file with updated listener addresses. This script is currently called
+# in the hot restart integration test to update listener addresses bound to
+# port 0 in the intial json config file.
+
+from collections import OrderedDict
+
+import argparse
+import httplib
+import json
+import os.path
+import sys
+import time
+
+# Seconds to wait for the admin address output file to appear. The script exits
+# if the file is not found.
+ADMIN_FILE_TIMEOUT_SECS = 20
+
+def ReplaceListenerAddresses(original_json, admin_address, updated_json):
+  # Get original listener addresses
+  with open(original_json, 'r') as original_json_file:
+    # Import original config file in order to get a deterministic output. This
+    # allows us to diff the original config file and the updated config file
+    # output from this script to check for any changes.
+    parsed_json = json.load(original_json_file, object_pairs_hook=OrderedDict)
+  original_listeners = parsed_json['listeners']
+
+  sys.stdout.write('Admin address is ' + admin_address + '\n')
+  try:
+    admin_conn = httplib.HTTPConnection(admin_address)
+    admin_conn.request('GET', '/listeners')
+    admin_response = admin_conn.getresponse()
+    if not admin_response.status == 200:
+      return False
+    discovered_listeners = json.loads(admin_response.read())
+  except Exception as e:
+    sys.stderr.write('Cannot connect to admin: %s\n' % e)
+    return False
+  else:
+    if not len(discovered_listeners) == len(original_listeners):
```

Comment:
`A != B` is clearer than `not A == B`. The second is in danger of being read `(not A) == B`, which isn't what it means, but the casual reader might easily make that mistake.